File: Java_ChatGPT/State/Round3/Code.java
package State.Round3;

// State interface
interface State {
    void turnOn(Light light);
    void turnOff(Light light);
}

// Concrete state for "On"
class OnState implements State {
    @Override
    public void turnOn(Light light) {
        System.out.println("The light is already on.");
    }

    @Override
    public void turnOff(Light light) {
        System.out.println("Turning off the light.");
        light.setState(new OffState());
    }
}

// Concrete state for "Off"
class OffState implements State {
    @Override
    public void turnOn(Light light) {
        System.out.println("Turning on the light.");
        light.setState(new OnState());
    }

    @Override
    public void turnOff(Light light) {
        System.out.println("The light is already off.");
    }
}

// Context class (Light)
class Light {
    private State state;

    public Light() {
        this.state = new OffState(); // Default state is off
    }

    public void setState(State state) {
        this.state = state;
    }

    public void turnOn() {
        state.turnOn(this);
    }

    public void turnOff() {
        state.turnOff(this);
    }

    public State getState() {
        return state;
    }
}

File: ChatGPT3_5/State/round1/Code.py
from abc import ABC, abstractmethod


# State interface
class State(ABC):
    @abstractmethod
    def turn_on(self, light):
        pass

    @abstractmethod
    def turn_off(self, light):
        pass


# Concrete state: On
class OnState(State):
    def turn_on(self, light):
        print("The light is already on.")

    def turn_off(self, light):
        print("Turning the light off.")
        light.set_state(OffState())


# Concrete state: Off
class OffState(State):
    def turn_on(self, light):
        print("Turning the light on.")
        light.set_state(OnState())

    def turn_off(self, light):
        print("The light is already off.")


# Context class: Light
class Light:
    def __init__(self):
        self.state = OffState()  # Default state is Off

    def set_state(self, state):
        self.state = state

    def turn_on(self):
        self.state.turn_on(self)

    def turn_off(self):
        self.state.turn_off(self)
